packages/core/src/components/forms/asyncControllableInput.tsx

```diff
@@ -134,17 +142,21 @@ export class AsyncControllableInput extends React.PureComponent<
     }
 
     private handleCompositionStart = (e: React.CompositionEvent<HTMLInputElement>) => {
-        this.setState({
-            isComposing: true,
-            // Make sure that localValue matches externalValue, in case externalValue
-            // has changed since the last onChange event.
-            nextValue: this.state.value,
-        });
+        this.cancelPendingCompositionEnd?.();
+        this.setState({ isComposing: true });
         this.props.onCompositionStart?.(e);
     };
 
     private handleCompositionEnd = (e: React.CompositionEvent<HTMLInputElement>) => {
-        this.setState({ isComposing: false });
+        // In some non-latin languages, a keystroke can end a composition event and immediately afterwards start another.
+        // This can lead to unexpected characters showing up in the text input. In order to circumvent this problem, we
+        // use a timeout which creates a delay which merges the two composition events, creating a more natural and predictable UX.
+        // `this.state.nextValue` will become "locked" (it cannot be overwritten by the `value` prop) until a delay (10ms) has
+        // passed without a new composition event starting.
+        this.cancelPendingCompositionEnd = this.setTimeout(
+            () => this.setState({ isComposing: false }),
+            AsyncControllableInput.COMPOSITION_END_DELAY,
+        );
         this.props.onCompositionEnd?.(e);
```

Comment:
@Athenodoros looks like we still fire the `onCompositionEnd` callback without the new delay... that sounds fine to me for now, but I just want to make sure that doesn't cause any problems for you using the component with korean characters?

Reply:
I think it's fine - anyone overriding onCompositionEnd should probably have the native behaviour (and I think React's synthetic events don't work well with timeouts anyway?). If it becomes a problem then I think they're probably doing something specific enough that they can manage the composition events themselves...